File: benchmark/sieve.go
// Prime Sieve of Eratosthenes - Go Implementation
package main

import (
	"fmt"
	"time"
)

func sieve(n int64) int64 {
	// Create sieve slice (true = potentially prime)
	isPrime := make([]bool, n+1)
	for i := range isPrime {
		isPrime[i] = true
	}
	isPrime[0] = false
	isPrime[1] = false

	// Sieve of Eratosthenes
	for i := int64(2); i*i <= n; i++ {
		if isPrime[i] {
			// Mark multiples as composite
			for j := i * i; j <= n; j += i {
				isPrime[j] = false
			}
		}
	}

	// Count primes
	var count int64 = 0
	for i := int64(2); i <= n; i++ {
		if isPrime[i] {
			count++
		}
	}
	return count
}

func main() {
	fmt.Println("Prime Sieve Benchmark")
	fmt.Println("=====================")

	sizes := []int64{100000, 500000, 1000000, 5000000, 10000000}

	for _, n := range sizes {
		fmt.Printf("Sieving up to: %d\n", n)

		start := time.Now()
		count := sieve(n)
		elapsed := time.Since(start)

		fmt.Printf("Primes found: %d\n", count)
		fmt.Printf("Time: %.2f ms\n\n", float64(elapsed.Nanoseconds())/1e6)
	}

	fmt.Println("Benchmark complete!")
}

File: benchmark/sieve.rs
// Prime Sieve of Eratosthenes - Rust Implementation
use std::time::Instant;

fn sieve(n: i64) -> i64 {
    // Create sieve vector (true = potentially prime)
    let mut is_prime = vec![true; (n + 1) as usize];
    is_prime[0] = false;
    is_prime[1] = false;
    
    // Sieve of Eratosthenes
    let sqrt_n = (n as f64).sqrt() as i64;
    for i in 2..=sqrt_n {
        if is_prime[i as usize] {
            // Mark multiples as composite
            let mut j = i * i;
            while j <= n {
                is_prime[j as usize] = false;
                j += i;
            }
        }
    }
    
    // Count primes
    is_prime.iter().filter(|&&p| p).count() as i64
}

fn main() {
    println!("Prime Sieve Benchmark");
    println!("=====================");
    
    let sizes = [100000, 500000, 1000000, 5000000, 10000000];
    
    for &n in &sizes {
        println!("Sieving up to: {}", n);
        
        let start = Instant::now();
        let count = sieve(n);
        let elapsed = start.elapsed();
        
        println!("Primes found: {}", count);
        println!("Time: {:.2} ms", elapsed.as_secs_f64() * 1000.0);
        println!();
    }
    
    println!("Benchmark complete!");
}
